src/components/NostoCampaign.ts

```diff
@@ -0,0 +1,55 @@
+import { assertRequired } from "@/utils"
+import { customElement } from "./decorators"
+import { nostojs } from "@nosto/nosto-js"
+
+@customElement("nosto-campaign")
+export class NostoCampaign extends HTMLElement {
+  static attributes = {
+    placement: String,
+    product: String,
+    variant: String
+  }
+
+  placement?: string
+  product?: string
+  variant?: string
+
+  async connectedCallback() {
+    assertRequired(this, "placement")
+    await this.loadCampaign()
+  }
+
+  async loadCampaign() {
+    const api = await new Promise(nostojs)
+    const request = api
+      .createRecommendationRequest({ includeTagging: true })
+      .disableCampaignInjection()
+      .setElements([this.placement!])
+      .setResponseMode("HTML")
+
+    if (this.product) {
+      request.setProducts([
+        {
+          product_id: this.product,
+          ...(this.variant ? { sku_id: this.variant } : {})
+        }
+      ])
+    }
+
+    const result = await request.load()
+    const rec = result.recommendations[this.placement!]
+
+    const html =
+      typeof rec === "string"
+        ? rec
+        : typeof rec === "object" && rec !== null && "html" in rec
+          ? (rec as { html: string }).html
+          : undefined
+
+    if (html) {
+      this.innerHTML = html
```

Comment:
direct assignment of the HTML has the problem that this fully skips the CSS and script handling of the client script
also attribution handling is missing